### PR TITLE
Weight pruning and visualization of weight mask / pruned weights

### PR DESCRIPTION
**Pull request by Hrancheng**
Added one argument "--prune", to enable model pruning after training and accuracy computation in train.py

Added arguments in sample.py, "--prune_percentage" to specify the percentage of weights to prune, and reused previous code to visualize weights/weight mask:

Pruned weights:
layer_0_weights (4)

Weight mask: 
transformer h 0 mlp c_proj_weight_mask

### CODE REVIEW

File: sample.py [hunk 1]
```diff
@@ -38,7 +40,9 @@ def parse_args():
     parser.add_argument('--block_size', type=int, default=None, help="Block size for context length, default is model's block size")
     parser.add_argument('--sym_rot_num_angles', type=int, default=None, help="Number of angles for symmetrical rotary embedding")
     parser.add_argument('--token_boundary', type=str, default=None, help="optional separator between emitted tokens")
-
+    parser.add_argument("--quant_weights_file", type=str, default=None, help="File to export the quantized weights and scale factor")
```
Comment (gkielian) [inline sample.py:43]: Great! This is something we need for interfacing with numpy, would this be in npy format?

File: sample.py [hunk 2]
```diff
@@ -38,7 +40,9 @@ def parse_args():
...
+    parser.add_argument("--visualize_weights_dir", type=str, default=None, help="Folder to save heatmaps of attention weights for all layers")
```
Comment (gkielian) [inline sample.py:44]: Maybe default to "visualizations"

File: sample.py [hunk 3]
```diff
@@ -38,7 +40,9 @@ def parse_args():
...
+    parser.add_argument("--prune_percentage", type=float, default=0, help="Percentage of weights to prune")
```
Comment (gkielian) [inline sample.py:45]: Let's make default None and handle with if else blocks.

File: sample.py [hunk 4]
```diff
@@ -88,7 +92,73 @@ def save_args(args, out_dir):
...
+    with open(filename, 'rb') as f:
+        weights = pickle.load(f)
+    for i in range(n_layers):
+        plt.rcParams["figure.figsize"] = [11, 3.5]
+        plt.rcParams["figure.autolayout"] = True
+        fig, (ax1, ax2, ax3) = plt.subplots(ncols=3)
+        if f"transformer.h.{i}.attn.c_attn_q.weight" in weights:
+            q_key = f"transformer.h.{i}.attn.c_attn_q.weight"
+            k_key = f"transformer.h.{i}.attn.c_attn_k.weight"
+            v_key = f"transformer.h.{i}.attn.c_attn_v.weight"
```
Comment (gkielian) [inline sample.py:116]: Add word table embedding weights

Add projection linear as well.

RMSnorm gain weights (for attention and output layer)

Finally MLP linear transform weights (for now let's start with the standard up projection and downprojection style default MLP setting)

With these we should be able to create a test in numpy.

File: sample.py [hunk 5]
```diff
@@ -88,7 +92,73 @@ def save_args(args, out_dir):
...
+        sns.heatmap(weights[q_key], ax=ax1)
+        sns.heatmap(weights[k_key], ax=ax2)
+        sns.heatmap(weights[v_key], ax=ax3)
+        ax1.set_title(f"Heatmap of Query Weights for Layer {i}")
+        ax2.set_title(f"Heatmap of Key Weights for Layer {i}")
+        ax3.set_title(f"Heatmap of Value Weights for Layer {i}")
+        #save to local dir
+        #create a dir if it does not exist
+        os.makedirs(weights_dir, exist_ok=True)
+        plt.savefig(f"{weights_dir}/layer_{i}_weights.png")
```
Comment (gkielian) [inline sample.py:126]: Let's have one for embedding table (WTE) 
One for QKV
one for Attention projection
One for MLP Up/Down projections
and one for the gains from RMSNorms.

File: sample.py [hunk 6]
```diff
@@ -88,7 +92,73 @@ def save_args(args, out_dir):
...
+
+def apply_pruning(model, prune_percentage, visualize_weights_dir):
```
Comment (gkielian) [inline sample.py:128]: We should create a new folder for training variations maybe call it "pruning" and let's place this function there.

This will help reduce the size of the sample.py file

File: train.py [hunk 7]
```diff
@@ -728,6 +731,21 @@ def plot_statistics(self, graph_y_labels):
...
 
+    def apply_pruning(self, prune_percentage):
+        parameters_to_prune = []
+        for name, module in self.model.named_modules():
+            if isinstance(module, torch.nn.Linear):
+                parameters_to_prune.append((module, 'weight'))
+
+        prune.global_unstructured(
+            parameters_to_prune,
+            pruning_method=prune.L1Unstructured,
```
Comment (gkielian) [inline train.py:742]: What other methods are there? Please add a README.md to the pruning/ directory briefly discussing maybe top 3.

Also add argparse options for these.

File: train.py [hunk 8]
```diff
@@ -1015,6 +1033,14 @@ def train(self):
...
             # End of training actions
             if self.iter_num > self.args.max_iters:
+                if self.args.prune:
+                    original_model = self.model
+                    for prune_percentage in [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]:
+                        self.model = original_model
+                        self.apply_pruning(prune_percentage)
+                        losses = self.estimate_loss()
+                        print(f"Prune Percentage: {prune_percentage}, val loss {losses['val']:.4f}")
+                    self.model = original_model
```
Comment (gkielian) [inline train.py:1043]: Would be interested in sending these statistics to tensorboard too, optionally, at the end of the validation loss measurement.

**Review by gkielian — CHANGES_REQUESTED** [review]
Looks good, requesting some changes to
